Handle sporadic bug in Android's security library

Google appear to be in no rush to fix [a bug](https://issuetracker.google.com/issues/164901843) that's causing exceptions like this:

```
FATAL EXCEPTION: DefaultDispatcher-worker-4
Process: tech.relaycorp.letro, PID: 32490
javax.crypto.AEADBadTagException
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:611)
	at javax.crypto.Cipher.doFinal(Cipher.java:2113)
	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decryptInternal(AndroidKeystoreAesGcm.java:118)
	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decrypt(AndroidKeystoreAesGcm.java:101)
	at com.google.crypto.tink.KeysetHandle.decrypt(KeysetHandle.java:919)
	at com.google.crypto.tink.KeysetHandle.readWithAssociatedData(KeysetHandle.java:804)
	at com.google.crypto.tink.KeysetHandle.read(KeysetHandle.java:785)
	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readMasterkeyDecryptAndParseKeyset(AndroidKeysetManager.java:381)
	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.build(AndroidKeysetManager.java:297)
	at androidx.security.crypto.EncryptedFile$Builder.build(EncryptedFile.java:233)
	at tech.relaycorp.awaladroid.AndroidPrivateKeyStore.buildEncryptedFile(AndroidPrivateKeyStore.kt:24)
	at tech.relaycorp.awaladroid.AndroidPrivateKeyStore.makeEncryptedInputStream(AndroidPrivateKeyStore.kt:14)
	at tech.relaycorp.awaladroid.AndroidPrivateKeyStore.makeEncryptedInputStream(AndroidPrivateKeyStore.kt:10)
	at tech.relaycorp.awala.keystores.file.FilePrivateKeyStore.retrieveKeyData(FilePrivateKeyStore.kt:73)
	at tech.relaycorp.awala.keystores.file.FilePrivateKeyStore.retrieveAllIdentityKeyData$suspendImpl(FilePrivateKeyStore.kt:31)
	at tech.relaycorp.awala.keystores.file.FilePrivateKeyStore.retrieveAllIdentityKeyData(Unknown Source:0)
	at tech.relaycorp.relaynet.keystores.PrivateKeyStore.retrieveAllIdentityKeys(PrivateKeyStore.kt:38)
	at tech.relaycorp.awaladroid.endpoint.RenewExpiringCertificates.invoke(RenewExpiringCertificates.kt:15)
	at tech.relaycorp.awaladroid.Awala$setUp$4$1.invokeSuspend(Awala.kt:59)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@ca9c6ff, Dispatchers.IO]
Caused by: android.security.KeyStoreException: Signature/MAC verification failed
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:356)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
	at android.security.KeyStoreOperation.finish(KeyStoreOperation.java:127)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.finish(KeyStoreCryptoOperationChunkedStreamer.java:228)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:181)
	at android.security.keystore2.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal(AndroidKeyStoreAuthenticatedAESCipherSpi.java:396)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:603)
	... 26 more
```

So I think we should catch `AEADBadTagException` and wrap it around another exception.

--- a/lib/src/main/java/tech/relaycorp/awaladroid/AndroidPrivateKeyStore.kt
+++ b/lib/src/main/java/tech/relaycorp/awaladroid/AndroidPrivateKeyStore.kt
@@ -11,6 +11,14 @@ import javax.crypto.AEADBadTagException
 internal class AndroidPrivateKeyStore(
     root: FileKeystoreRoot,
     private val context: Context,
+    private val encryptedFileBuilder: (File, MasterKey) -> EncryptedFile = { file, masterKey ->
+        EncryptedFile.Builder(
+            context,
+            file,
+            masterKey,
+            EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB,
+        ).build()
+    },
 ) : FilePrivateKeyStore(root) {
 
     @Throws(EncryptionInitializationException::class)
@@ -20,17 +28,15 @@ internal class AndroidPrivateKeyStore(
     override fun makeEncryptedOutputStream(file: File) = buildEncryptedFile(file).openFileOutput()
 
     @Throws(EncryptionInitializationException::class)
-    private fun buildEncryptedFile(file: File) =
+    private fun buildEncryptedFile(file: File): EncryptedFile =
         try {
-            EncryptedFile.Builder(
-                context,
-                file,
-                masterKey,
-                EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB,
-            ).build()
-        } catch (e: AEADBadTagException) {
+            encryptedFileBuilder(file, masterKey)
+        } catch (exception: AEADBadTagException) {
             // Known issue: https://issuetracker.google.com/issues/164901843
-            throw EncryptionInitializationException(e)
+            throw EncryptionInitializationException(
+                "Could not build encrypted file due to internal issue",
+                exception,
+            )
         }
 
     private val masterKey by lazy {
@@ -39,9 +45,10 @@ internal class AndroidPrivateKeyStore(
             .build()
     }
 
-    class EncryptionInitializationException(cause: Throwable) : Exception(cause)
-
     companion object {
         private const val MASTER_KEY_ALIAS = "_awaladroid_master_key_"
     }
 }
+
+public class EncryptionInitializationException(message: String, cause: Throwable) :
+    AwaladroidException(message, cause)

--- a/lib/src/main/java/tech/relaycorp/awaladroid/AndroidPrivateKeyStore.kt
+++ b/lib/src/main/java/tech/relaycorp/awaladroid/AndroidPrivateKeyStore.kt
@@ -6,28 +6,40 @@ import androidx.security.crypto.MasterKey
 import tech.relaycorp.awala.keystores.file.FileKeystoreRoot
 import tech.relaycorp.awala.keystores.file.FilePrivateKeyStore
 import java.io.File
+import javax.crypto.AEADBadTagException
 
 internal class AndroidPrivateKeyStore(
     root: FileKeystoreRoot,
     private val context: Context,
 ) : FilePrivateKeyStore(root) {
+
+    @Throws(EncryptionInitializationException::class)
     override fun makeEncryptedInputStream(file: File) = buildEncryptedFile(file).openFileInput()
 
+    @Throws(EncryptionInitializationException::class)
     override fun makeEncryptedOutputStream(file: File) = buildEncryptedFile(file).openFileOutput()
 
+    @Throws(EncryptionInitializationException::class)
     private fun buildEncryptedFile(file: File) =
-        EncryptedFile.Builder(
-            context,
-            file,
-            masterKey,
-            EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB,
-        ).build()
+        try {
+            EncryptedFile.Builder(
+                context,
+                file,
+                masterKey,
+                EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB,
+            ).build()
+        } catch (e: AEADBadTagException) {
+            // Known issue: https://issuetracker.google.com/issues/164901843
+            throw EncryptionInitializationException(e)
+        }
 
     private val masterKey by lazy {
         MasterKey.Builder(context, MASTER_KEY_ALIAS)
             .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
             .build()
     }
+
+    class EncryptionInitializationException(cause: Throwable) : Exception(cause)
 
     companion object {
         private const val MASTER_KEY_ALIAS = "_awaladroid_master_key_"

--- a/lib/src/test/java/tech/relaycorp/awaladroid/AndroidPrivateKeyStoreTest.kt
+++ b/lib/src/test/java/tech/relaycorp/awaladroid/AndroidPrivateKeyStoreTest.kt
@@ -12,6 +12,7 @@ import tech.relaycorp.awaladroid.test.FakeAndroidKeyStore
 import tech.relaycorp.relaynet.testing.pki.KeyPairSet
 import tech.relaycorp.relaynet.testing.pki.PDACertPath
 import java.io.File
+import javax.crypto.AEADBadTagException
 
 @RunWith(RobolectricTestRunner::class)
 public class AndroidPrivateKeyStoreTest {
@@ -32,5 +33,15 @@ public class AndroidPrivateKeyStoreTest {
         store.saveIdentityKey(id)
         val retrievedId = store.retrieveIdentityKey(certificate.subjectId)
         assertEquals(id, retrievedId)
+    }
+
+    @Test(expected = EncryptionInitializationException::class)
+    public fun failWithAEADBadTagException(): Unit = runTest {
+        val androidContext = RuntimeEnvironment.getApplication()
+        val root = FileKeystoreRoot(File(androidContext.filesDir, "tmp-keystore"))
+        val store = AndroidPrivateKeyStore(root, androidContext) { _, _ ->
+            throw AEADBadTagException("")
+        }
+        store.saveIdentityKey(KeyPairSet.PRIVATE_ENDPOINT.private)
     }
 }

